### ISSUE TITLE
Consider adding OSC 133

### ISSUE THREAD

**Issue by kchibisov** [alacritty/alacritty#5850]
This allows some basic shell integration and the protocol seems like supported by some major modern terminals (kitty/iterm2).

Adding it should be relatively easy and shouldn't impact performance at all.

I'm not sure that adding all of them makes sense for us, but ability to just mark prompt should be fine.

The relevant spec could be found https://iterm2.com/documentation-escape-codes.html.

**Comment by kchibisov** [comment]
Also, I think it should be restricted in alt-screen?

**Comment by chrisduerr** [comment]
> Also, I think it should be restricted in alt-screen?

would just be a no-op in alt screen, wouldn't it?

**Comment by kchibisov** [comment]
I don't think that adding this escape sequence will provide any value due ability to implemented it yourself with bindings due to 586f982.

I wouldn't close this issue out right, but I don't think this escape sequence will add anything our macro like bindings can't do. Especially given that the escape sequence is unclear an complicated.

There's a part of the escape sequence talking about last prompt copy, but it's useless due to `!!` shell builtin for more you can arrow cycle.

**Comment by nixpulvis** [comment]
Curious, as a developer of an application/shell, how would this make my life easier better? Let's say I emit the correct starts and ends for commands run by the user. What actions would I be able to do with this? Do other applications look for these escapes? Would we track them for our own internal tracking of commands (seems wrong).

When I've played with writing shells in the past, I've considered adding command tracking (even in parallel with `jobs` support), but I'm not sure what terminal features are really needed.

The killer feature is probably "Jump to Last Command", which if I understand you correctly @kchibisov you think should be implementable with out existing bindings? What would that look like without this OSC?

Hopefully I'm thinking along the right lines here. It's been a little while since I gave this much thought.

**Comment by kchibisov** [comment]
>Curious, as a developer of an application/shell, how would this make my life easier better? Let's say I emit the correct starts and ends for commands run by the user. What actions would I be able to do with this? Do other applications look for these escapes? Would we track them for our own internal tracking of commands (seems wrong).

it's only for the terminal itself.

>The killer feature is probably "Jump to Last Command", which if I understand you correctly @kchibisov you think should be implementable with out existing bindings? What would that look like without this OSC?

```
  { key = "P",      mods = "Control|Shift", mode = "~Vi|~Search",  action = "ToggleViMode" },
  { key = "P",      mods = "Control|Shift", mode = "~Vi|~Search",  action = "SearchBackward" },
  { key = "P",      mods = "Control|Shift", mode = "~Vi|~Search",  chars = " " },
  { key = "P",      mods = "Control|Shift", mode = "~Vi|~Search",  action = "SearchConfirm" },
```

**Comment by nixpulvis** [comment]
I can’t tell at the moment, is that `chars` searching for a specific invisible character?

**Comment by kchibisov** [comment]
yeah, it's a special space char.

**Comment by halvabner** [comment]
Hello, looks like `PreviousCommand` and other actions were added to the commit but now they're no longer there?
I am getting `[alacritty_config_derive] Config error: unknown keyboard action `PreviousPrompt`; ignoring binding` when I try to use them

I tried @kchibisov 's solution (had to change some things to make it parseable)
```yaml
  - { key: P,    mods: Control|Shift,    mode: ~Vi|~Search,  action: ToggleViMode }
  - { key: P,    mods: Control|Shift,    mode: ~Vi|~Search,  action: SearchBackward }
  - { key: P,    mods: Control|Shift,    mode: ~Vi|~Search,  chars: "test" }
  - { key: P,    mods: Control|Shift,    mode: Search,  action: SearchConfirm }  # errors if ~Vi|~Search
```
but the `chars` won't get written in the search bar

**Comment by kchibisov** [comment]
@halvabner you need master built to make it work.

**Comment by halvabner** [comment]
@kchibisov sorry, can you elaborate please? Are you suggesting I am not running a master branch build? I have `alacritty 0.12.2 (9d9982df)` using [this build](https://gitlab.archlinux.org/archlinux/packaging/packages/alacritty/-/blob/main/PKGBUILD?ref_type=heads) when I check. Where are the build instructions to build the version you refer to?

**Comment by kchibisov** [comment]
if you're on arch, you need alacritty-git from aur. The instructions on how to build yourself are in INSTALL.md.

**Comment by halvabner** [comment]
@kchibisov and how do you specify OSC character? I tried
```
[[keyboard.bindings]]
chars = ']133;A\'
key = "P"
mode = "~Vi|~Search"
mods = "Control|Shift"
```
which won't get parsed. I also tried writing bytes
![image](https://github.com/alacritty/alacritty/assets/130675714/ef60da73-570d-4868-9633-da2b0f1499ae)

**Comment by kchibisov** [comment]
They are not part of the grid. Just use some unique character in your prompt, like some unicode variant of space or some arrow character you could have.

Besides, TOML uses `\u` for unicode input.

**Comment by theimpostor** [comment]
The 'search mode trick' doesn't seem to work for SearchForward. I am using the following config block on mac to use <kbd>⌘ Command</kbd> + <kbd>↑</kbd> and <kbd>⌘ Command</kbd> + <kbd>↓</kbd> to search for a special prompt character (`❯`):

```toml
[keyboard]
bindings = [
# Use cmd-up and cmd-down to jump to previous/next prompt in scrollback
{ key = "ArrowUp",   mods = "Command",      mode = "~Vi|~Search", action = "ToggleViMode" },
{ key = "ArrowUp",   mods = "Command",      mode = "~Vi|~Search", action = "SearchBackward" },
{ key = "ArrowUp",   mods = "Command",      mode = "~Vi|~Search", chars = "❯" },
{ key = "ArrowUp",   mods = "Command",      mode = "~Vi|~Search", action = "SearchConfirm" },
{ key = "ArrowUp",   mods = "Command",      mode = "Vi",          action = "SearchBackward" },
{ key = "ArrowUp",   mods = "Command",      mode = "Vi",          chars = "❯" },
{ key = "ArrowUp",   mods = "Command",      mode = "Vi",          action = "SearchConfirm" },
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", action = "ToggleViMode" },
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", action = "SearchForward" },
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", chars = "❯" },
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", action = "SearchConfirm" },
{ key = "ArrowDown", mods = "Command",      mode = "Vi",          action = "SearchForward" },
{ key = "ArrowDown", mods = "Command",      mode = "Vi",          chars = "❯" },
{ key = "ArrowDown", mods = "Command",      mode = "Vi",          action = "SearchConfirm" },
]
```

<kbd>⌘ Command</kbd> + <kbd>↑</kbd> appears to work and can be repeated to jump further and further back in history. However the <kbd>⌘ Command</kbd> + <kbd>↓</kbd> does not seem to work at all. Any ideas?

**Comment by chrisduerr** [comment]
@theimpostor Does it work when done manually? Multi-bindings should effectively be the same as pressing them in sequence, so if it works manually it should also work combined.

**Comment by theimpostor** [comment]
Yes if I manually enter Vi mode, do <kbd>/</kbd> and search for the character it works

**Comment by chrisduerr** [comment]
It works on my machine.

**Comment by haarp** [comment]
Can confirm @theimpostor. `SearchBackward` works fine, while `SearchForward` seems broken. It does weird stuff, like search backwards, but only sometimes, and eventually stops working entirely.

**Comment by gandalfs-cat** [comment]
> The search mode trick doesn't seem to work for `SearchForward`. I am using the following config block on mac to use ⌘ Command + ↑ and ⌘ Command + ↓ to search for a special prompt character '❯':

for me this is because `SearchForward` doesn't move the cursor to a new line when initiated on a line with a match; it even goes backwards if there's a match on the preceding line! adding two `ScrollLineDown`s like this made it work:

```toml
[keyboard]
# Use cmd-up and cmd-down to jump to previous/next prompt in scrollback
bindings = [
{ key = "ArrowUp",   mods = "Command",      mode = "~Vi|~Search", action = "ToggleViMode"  },
{ key = "ArrowUp",   mods = "Command",      mode = "~Vi|~Search", action = "SearchBackward"},
{ key = "ArrowUp",   mods = "Command",      mode = "~Vi|~Search", chars = "❯" },
{ key = "ArrowUp",   mods = "Command",      mode = "~Vi|~Search", action = "SearchConfirm" },
{ key = "ArrowUp",   mods = "Command",      mode = "Vi",          action = "SearchBackward"},
{ key = "ArrowUp",   mods = "Command",      mode = "Vi",          chars = "❯" },
{ key = "ArrowUp",   mods = "Command",      mode = "Vi",          action = "SearchConfirm" },
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", action = "ToggleViMode"  },
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", action = "ScrollLineDown"}, # scroll to get SearchForward to move
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", action = "ScrollLineDown"},
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", action = "SearchForward" },
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", chars = "❯" },
{ key = "ArrowDown", mods = "Command",      mode = "~Vi|~Search", action = "SearchConfirm" },
{ key = "ArrowDown", mods = "Command",      mode = "Vi",          action = "ScrollLineDown"}, # scroll to get SearchForward to move
{ key = "ArrowDown", mods = "Command",      mode = "Vi",          action = "ScrollLineDown"},
{ key = "ArrowDown", mods = "Command",      mode = "Vi",          action = "SearchForward" },
{ key = "ArrowDown", mods = "Command",      mode = "Vi",          chars = "❯" },
{ key = "ArrowDown", mods = "Command",      mode = "Vi",          action = "SearchConfirm" },
]
```

i also made a version which skips prompts already in view and doesn't end in vi mode (i found this more intuitive for me):
<details><summary>expand snippet</summary>

```toml
[keyboard]
# Use ctrl-up and ctrl-down to jump to previous/next prompt in scrollback
bindings = [
{ key = "ArrowUp",   mods = "Control",      mode = "~Vi|~Search", action = "ToggleViMode" },
{ key = "ArrowUp",   mods = "Control",      mode = "~Vi|~Search", action = "ScrollPageUp" },
{ key = "ArrowUp",   mods = "Control",      mode = "~Vi|~Search", action = "SearchBackward" },
{ key = "ArrowUp",   mods = "Control",      mode = "~Vi|~Search", chars = "❯" },
{ key = "ArrowUp",   mods = "Control",      mode = "~Vi|~Search", action = "SearchConfirm" },
{ key = "ArrowUp",   mods = "Control",      mode = "~Vi|~Search", action = "CenterAroundViCursor" },
{ key = "ArrowUp",   mods = "Control",      mode = "~Vi|~Search", action = "ToggleViMode" },

{ key = "ArrowDown", mods = "Control",      mode = "~Vi|~Search", action = "ToggleViMode" },
{ key = "ArrowDown", mods = "Control",      mode = "~Vi|~Search", action = "ScrollPageDown" },
{ key = "ArrowDown", mods = "Control",      mode = "~Vi|~Search", action = "SearchForward" },
{ key = "ArrowDown", mods = "Control",      mode = "~Vi|~Search", chars = "❯" },
{ key = "ArrowDown", mods = "Control",      mode = "~Vi|~Search", action = "SearchConfirm" },
{ key = "ArrowDown", mods = "Control",      mode = "~Vi|~Search", action = "CenterAroundViCursor" },
{ key = "ArrowDown", mods = "Control",      mode = "~Vi|~Search", action = "ToggleViMode" },
]
```

</details>

but with the downside that the prompt is centered instead of aligned at the top; i didn't find a good way to achieve that consistently.

**Comment by nixpulvis** [comment]
I've created https://github.com/alacritty/vte/pull/152 to add support for OSC 133 to the VTE parser. This seems likely to be merged, so the next question is how `alacritty_terminal` should consume this.

My draft PR which prompted this (pun intended) was actually to support `fish_handle_reflow`: https://github.com/alacritty/alacritty/pull/8961, but A) I'm not happy with the prompt marking on the `Term` and B) it doesn't track the complete set of OSC 133 markers.

I'm thinking a better design would be to store the markers in a `Cell` with `Flags` and `CellExtra`. We seem to have exactly one bit left in `Flags` so it should fit without extending it, though there's a real question there of what to do when we want another new flag down the road. Saving this bit could be useful... So I'm open to suggestions here. `CellExtra` would store the OSC 133 marker kind.

If anyone has a thought on a better way to do this, I'm all ears. This avoids having to manually reflow marker points on resize and scroll though, which will be much more important when we start tracking not just the current prompt like my draft PR does, but every marker.
